### PR TITLE
wp-cli support to index by date range

### PR DESCRIPTION
**Pull request by alexisbellido**

### CODE REVIEW

**Comment by xmattus** [comment]
This seems good to me, but please look into the travis errors and make sure all the tests pass before merging. It looks like this isn't related to the actual functionality here, but has to do with some missing unrelated WP-API libraries. If you're stuck, ping me and we'll look at it. We just need to make sure Travis is working so we diagnose legitimately non-passing tests in the future.

**Comment by mboynes** [comment]
Please make it such that you can specify the start _or_ end and not have to specify both. I will definitely want to use this to index all posts after a certain date, and there should be no need to have to specify a future date as the end

File: bin/wp-cli.php [hunk 1]
```diff
@@ -171,6 +196,14 @@ public function index( $args, $assoc_args ) {
 			if ( $assoc_args['limit'] && $assoc_args['limit'] < $assoc_args['bulk'] )
 				$assoc_args['bulk'] = $assoc_args['limit'];
 
+			if ( isset( $assoc_args['from'] ) && isset( $assoc_args['to'] ) ) {
+				$this->date_range = array(
```
Comment (mboynes) [inline bin/wp-cli.php]: Please declare this property in the class

File: bin/wp-cli.php [hunk 2]
```diff
@@ -189,12 +222,14 @@ public function index( $args, $assoc_args ) {
...
+				if ( 0 < ( $sync_meta->page - $start_page ) ) {
+					$seconds_per_page = ( microtime( true ) - $timestamp_start ) / ( $sync_meta->page - $start_page );
+					WP_CLI::line( "Completed page {$sync_meta->page}/{$total_pages_ceil} (" . number_format( ( microtime( true ) - $lap), 2 ) . 's / ' . round( memory_get_usage() / 1024 / 1024, 2 ) . 'M current / ' . round( memory_get_peak_usage() / 1024 / 1024, 2 ) . 'M max), ' . $this->time_format( ( $total_pages - $sync_meta->page ) * $seconds_per_page ) . ' remaining' );
+				}
 
 				$this->contain_memory_leaks();
 
-				if ( $assoc_args['limit'] > 0 && $sync_meta->processed >= $assoc_args['limit'] )
+				if ( ( $assoc_args['limit'] > 0 && $sync_meta->processed >= $assoc_args['limit'] ) || 0 == $sync_meta->total )
 					break;
```
Comment (mboynes) [inline bin/wp-cli.php]: Coding standards update: please add curly braces around this

File: bin/wp-cli.php [hunk 3]
```diff
@@ -82,6 +82,31 @@ public function flush() {
...
 	}
 
+	/**
+	 * Add date range when retrieving posts in bulk.
+	 * Dates need to be passed as mmddyyy. See synopsis for index function.
+	 *
+	 * @param $args array
+	 * @return $args array
+	 */
+	public function apply_date_range( $args ) {
```
Comment (mboynes) [inline bin/wp-cli.php]: WP-CLI may see this as a subcommand because it's public. Can you verify that it doesn't (`wp help searchpress`)? If it does, I'm sure there's a phpdoc flag to indicate that it isn't.

Comment (alexisbellido) [inline bin/wp-cli.php]: You're right. WP-CLI sees apply_date_range as a subcommand and if I make it private the add_filter call won't work. I haven't found a way to tell phpdoc to ignore a public method.

Jetpack works around this by putting a function outside the class (https://github.com/Automattic/jetpack/blob/master/class.jetpack-cli.php#L573) but in this case I'm using add_filter and by using a function instead of a method I lose the ability to pass $this->date_range.

I could pass the date range with a global variable but I don't feel that's too elegant, see https://github.com/alleyinteractive/searchpress/commit/a93e4c2d3800ad6fd61f0db9cd898b3672e70d72. I'm open to suggestions.

**Comment by alexisbellido** [comment]
Tests passing now. I borrowed the travis setup from Fieldmanager. I think I could just remove bin/install-wp-tests.sh as everything is setup in .travis.yml now.

**Comment by danielbachhuber** [comment]
> I think I could just remove bin/install-wp-tests.sh as everything is setup in .travis.yml now.

If you use `bin/install-wp-tests.sh`, you can more easily apply the myriad of improvements we continue to make to the script.

File: .travis.yml [hunk 4]
```diff
@@ -6,8 +6,12 @@ php:
...
-    - WP_VERSION=latest WP_MULTISITE=0
-    - WP_VERSION=latest WP_MULTISITE=1
+    - WP_VERSION=master WP_MULTISITE=0
+    - WP_VERSION=master WP_MULTISITE=1
+    - WP_VERSION=4.0 WP_MULTISITE=0
+    - WP_VERSION=4.0 WP_MULTISITE=1
+    - WP_VERSION=3.9 WP_MULTISITE=0
+    - WP_VERSION=3.9 WP_MULTISITE=1
     - WP_VERSION=3.8 WP_MULTISITE=0
     - WP_VERSION=3.8 WP_MULTISITE=1
```
Comment (mboynes) [inline .travis.yml]: these versions are all outdated. plus update this to 4.2, 4.3, and master (and remove the 4th version)

File: bin/wp-cli.php [hunk 5]
```diff
@@ -171,6 +175,18 @@ public function index( $args, $assoc_args ) {
 			if ( $assoc_args['limit'] && $assoc_args['limit'] < $assoc_args['bulk'] )
 				$assoc_args['bulk'] = $assoc_args['limit'];
 
+			if ( isset( $assoc_args['from'] ) || isset( $assoc_args['to'] ) ) {
+				global $searchpress_cli_date_range;
```
Comment (mboynes) [inline bin/wp-cli.php]: revert to how you were doing this before. move the info to a class property, and for your callback method for the filter, start the method name with two underscores. wp-cli will ignore all methods starting with `__` and won't turn them into subcommands.

File: bin/wp-cli.php [hunk 6]
```diff
@@ -128,8 +127,13 @@ public function flush() {
...
 	 *      wp searchpress index 12340 12341 12342 12343 12344 12345
 	 *
+	 *      # Index posts published between 11/1/2015 and 12/15/2015 inclusive
+	 *      wp searchpress index --from=11012015 --to=12152015
+	 *
+	 *      # Index posts published after 11/1/2015
+	 *      wp searchpress index --from=11012015
 	 *
-	 * @synopsis [--flush] [--put-mapping] [--bulk=<num>] [--limit=<num>] [--page=<num>] [<post-id>]
+	 * @synopsis [--flush] [--put-mapping] [--bulk=<num>] [--limit=<num>] [--page=<num>] [--from=<mmddyyyy>] [--to=<mmddyyyy>] [<post-id>]
```
Comment (mboynes) [inline bin/wp-cli.php]: You need to also document these (see above the `## EXAMPLES` section)

Comment (mboynes) [inline bin/wp-cli.php]: the standard format for passing dates around in wp-cli is `YYYY-MM-DD`. For the purposes of documentation, make this `[--from=<date>] [--to=<date>]` and in the documentation you're going to add, specify that it should be in `YYYY-MM-DD` format

**Comment by xmattus** [comment]
Looks like we get a bonus modernization of the SP testing framework out of this PR too. Nice work!

:horse:

File: tests/test-searching.php [hunk 7]
```diff
@@ -359,22 +359,19 @@ function test_query_paged_and_posts_per_page() {
...
+				'tag-c',
+				'tag-b',
+				'tag-a',
+				'tags-a-b-c',
+				'raw-html-code',
+				'simple-markup-test',
+				'embedded-video',
+				'contributor-post-approved',
+				'one-comment',
+				'no-comments',
```
Comment (mboynes) [inline tests/test-searching.php]: You need to revert the changes to this file. I'm sure there's a reason why you did this, so let's discuss in Slack, but this was very deliberate.

File: bin/wp-cli.php [hunk 8]
```diff
@@ -128,8 +165,13 @@ public function flush() {
 	 *      # Index six specific posts
 	 *      wp searchpress index 12340 12341 12342 12343 12344 12345
 	 *
+	 *      # Index posts published between 11-1-2015 and 12-15-2015 inclusive
+	 *      wp searchpress index --from=11-01-2015 --to=12-15-2015
```
Comment (mboynes) [inline bin/wp-cli.php]: See my previous comment that this should be in mysql date format, `YYYY-MM-DD`, to be consistent with other WP-CLI commands (as well as everything else in WP)

File: bin/wp-cli.php [hunk 9]
```diff
@@ -82,6 +84,35 @@ public function flush() {
...
+	 */
+	public function __apply_date_range( $args ) {
+		$args['date_query'] = array(
+			0 => array(
+				'inclusive' => true,
+			),
+		);
+		if ( isset ( $this->date_range['from'] ) ) {
+			$args['date_query'][0]['after'] = array(
+				'year'  => substr( $this->date_range['from'], -4 ),
```
Comment (mboynes) [inline bin/wp-cli.php]: Instead of using substr, you should use `strtotime()` and then use date functions to set these args. Also, are these inclusive? You should document whether or not they are.

File: bin/wp-cli.php [hunk 10]
```diff
@@ -82,6 +84,35 @@ public function flush() {
 		}
 	}
 
+ 	/**
+	 * Add date range when retrieving posts in bulk.
+	 * Dates need to be passed as mmddyyy. See synopsis for index function.
```
Comment (mboynes) [inline bin/wp-cli.php]: Update this

File: bin/wp-cli.php [hunk 11]
```diff
@@ -168,8 +210,23 @@ public function index( $args, $assoc_args ) {
...
+				if ( isset( $assoc_args['from'] ) ) {
+					$this->date_range['from'] = $assoc_args['from'];
+				}
+				if ( isset( $assoc_args['to'] ) ) {
+					$this->date_range['to'] = $assoc_args['to'];
+				}
+				add_filter( 'searchpress_index_loop_args', array( $this, '__apply_date_range' ) );
+				add_filter( 'searchpress_index_count_args', array( $this, '__apply_date_range' ) );
+
+
```
Comment (mboynes) [inline bin/wp-cli.php]: Remove these extra newlines

File: bin/wp-cli.php [hunk 12]
```diff
@@ -105,6 +136,12 @@ public function flush() {
 	 * : Which page to start on. This is helpful if you encountered an error on
 	 * page 145/150 or if you want to have multiple processes running at once
 	 *
+	 * [--from=<date>]
+	 * : Index posts published on or after this date.
+	 *
+	 * [--to=<date>]
+	 * : Index posts published on or before this date.
```
Comment (mboynes) [inline bin/wp-cli.php]: Specify the date format in these. These are parsed and output with `wp help`, so we want them to be as helpful as possible. Also, I made an earlier comment asking if the dates are inclusive. Looks like you specify that here, so disregard that!

File: bin/wp-cli.php (hunk 9 above)
Comment (mboynes) [inline bin/wp-cli.php]: Looks like they are inclusive and you did document that. Sorry!

File: bin/wp-cli.php [hunk 13]
```diff
@@ -128,8 +165,13 @@ public function flush() {
...
 	 *      wp searchpress index 12340 12341 12342 12343 12344 12345
 	 *
+	 *      # Index posts published between 11-1-2015 and 12-30-2015 inclusive
+	 *      wp searchpress index --from=2015-11-01 --to=2015-12-30
+	 *
+	 *      # Index posts published after 11-1-2015
+	 *      wp searchpress index --from=2015-11-01
 	 *
-	 * @synopsis [--flush] [--put-mapping] [--bulk=<num>] [--limit=<num>] [--page=<num>] [<post-id>]
+	 * @synopsis [--flush] [--put-mapping] [--bulk=<num>] [--limit=<num>] [--page=<num>] [--from=<date>] [--to=<date>] [<post-id>]
```
Comment (mboynes) [inline bin/wp-cli.php]: I wonder if "from" and "to" are the best arg names. Those words aren't date-oriented, they're range-oriented, and could imply dates, IDs, etc. What do you think about making them `--after-date` and `--before-date` instead? This would give us the ability to later do ID ranges without causing confusion.

**Comment by mboynes** [comment]
:100: Awesome work, thanks for adding this!
